File: storage/src/routes/versions.rs
//! Version-history endpoints for editable files.
//!
//! All operations are owner-only — there's no transfer-token surface
//! here, intentionally: history access is metadata + restore, neither of
//! which the upload/download proxy flow needs.

use actix_web::{route, web, HttpRequest, HttpResponse};
use auth::data::claims::Claims;
use chrono::Utc;
use context::Context;
use entity::{ActiveValue, TransactionTrait, Uuid};
use error::AppResult;
use fs::prelude::*;
use serde::{Deserialize, Serialize};

use crate::{
    data::create_file::CreateFile,
    repository::{cached::evict_file, Repository},
};

/// `GET /api/storage/{file_id}/versions` — newest-first list of
/// historical snapshots. The currently-active version is intentionally
/// absent (it lives on the file row, not in `file_versions`).
#[route("/api/storage/{file_id}/versions", method = "GET")]
pub(crate) async fn list(
    req: HttpRequest,
    claims: Claims,
    context: web::Data<Context>,
) -> AppResult<HttpResponse> {
    let context = context.into_inner();
    let file_id: Uuid = util::actix::path_var(&req, "file_id")?;

    let versions = Repository::new(&context.db)
        .versions(claims.sub)
        .list(file_id)
        .await?;

    Ok(HttpResponse::Ok().json(versions))
}

/// `GET /api/storage/{file_id}/versions/{version}` — fetch chunks of a
/// historical version. Mirrors the regular download endpoint:
/// `?format=tar` for a tar of all chunks, `?chunk=N` for a single chunk.
#[route("/api/storage/{file_id}/versions/{version}", method = "GET")]
pub(crate) async fn download(
    req: HttpRequest,
    claims: Claims,
    context: web::Data<Context>,
) -> AppResult<HttpResponse> {
    let context = context.into_inner();
    let file_id: Uuid = util::actix::path_var(&req, "file_id")?;
    let version: i32 = util::actix::path_var(&req, "version")?;
    let chunk = util::actix::query_var::<i64>(&req, "chunk").ok();
    let format = util::actix::query_var::<String>(&req, "format").ok();

    // Authorization + existence go through the repository — both file
    // and historical version must exist for the calling owner.
    let file = Repository::new(&context.db)
        .manage(claims.sub)
        .file(file_id)
        .await?;
    let _ = Repository::new(&context.db)
        .versions(claims.sub)
        .get(file_id, version)
        .await?;

    let storage = Fs::new(&context.config);

    if format.as_deref() == Some("tar") {
        let content_length = storage.tar_content_length_v(&file, version).await?;
        let streamer = storage.stream_tar_v(&file, version).await?;
        let filename = format!("{}.v{}.tar", file_id, version);

        return Ok(HttpResponse::Ok()
            .insert_header(("Content-Type", "application/x-tar"))
            .insert_header(("Content-Length", content_length.to_string()))
            .insert_header((
                "Content-Disposition",
                format!("attachment; filename=\"{filename}\""),
            ))
            .streaming(streamer.stream()));
    }

    let streamer = storage.stream_v(&file, version, chunk).await?;
    Ok(HttpResponse::Ok()
        .insert_header(("Content-Type", "application/octet-stream"))
        .streaming(streamer.stream()))
}

/// `POST /api/storage/{file_id}/versions/{version}/restore` —
/// in-place restore. Allocates a new version slot, copies the target's
/// chunks into it, and flips `active_version`. The previously-active
/// version is snapshotted into history along the way, so the user can
/// undo by restoring that one.
#[route(
    "/api/storage/{file_id}/versions/{version}/restore",
    method = "POST"
)]
pub(crate) async fn restore(
    req: HttpRequest,
    claims: Claims,
    context: web::Data<Context>,
) -> AppResult<HttpResponse> {
    let context = context.into_inner();
    let file_id: Uuid = util::actix::path_var(&req, "file_id")?;
    let version: i32 = util::actix::path_var(&req, "version")?;

    let storage = Fs::new(&context.config);

    let txn = context.db.begin().await?;
    let outcome = Repository::new(&txn)
        .versions(claims.sub)
        .restore(file_id, version)
        .await?;
    txn.commit().await?;

    // Pre-load the file via a fresh read so the chunks-copy below uses
    // the post-swap `IntoFilename` (timestamp/UUID unchanged, but it's
    // the right contract).
    let file = Repository::new(&context.db)
        .manage(claims.sub)
        .file(file_id)
        .await?;

    // Copy chunks out of band so the DB transaction stays small. If
    // copy fails the active pointer already moved — same recovery
    // story as a half-finished edit: the next purge_all on file delete
    // catches the dst dir, and a re-restore retries cleanly.
    storage
        .copy_version(&file, outcome.source_version, &file, outcome.new_version)
        .await?;

    evict_file(file_id).await;

    let file = Repository::new(&context.db)
        .manage(claims.sub)
        .file(file_id)
        .await?;
    Ok(HttpResponse::Ok().json(file))
}

/// Body for `fork`. Uses the `CreateFile` shape because the client
/// builds the new file's encrypted metadata exactly as it would for a
/// fresh upload — only the chunks are server-copied.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ForkRequest {
    /// Same fields as `createFile` — client-encrypted name/thumbnail,
    /// the file's symmetric key encrypted under the user's RSA pubkey,
    /// destination directory, etc. `chunks`/`size` are ignored here:
    /// the source version's recorded values are authoritative.
    #[serde(flatten)]
    pub file: CreateFile,
}

/// `POST /api/storage/{file_id}/versions/{version}/fork` —
/// restore-as-new-note. Creates a brand-new file row reusing the
/// source's chunks (same encryption key) and metadata supplied by the
/// client. The original file is untouched.
#[route("/api/storage/{file_id}/versions/{version}/fork", method = "POST")]
pub(crate) async fn fork(
    req: HttpRequest,
    claims: Claims,
    context: web::Data<Context>,
    data: web::Json<ForkRequest>,
) -> AppResult<HttpResponse> {
    let context = context.into_inner();
    let source_file_id: Uuid = util::actix::path_var(&req, "file_id")?;
    let source_version: i32 = util::actix::path_var(&req, "version")?;

    let source = Repository::new(&context.db)
        .manage(claims.sub)
        .file(source_file_id)
        .await?;

    // Source chunks/size/sha256 come from `file_versions` for historical
    // versions, or directly off the file row for the active one (no
    // history row exists for active until its first edit).
    let (chunks, size, sha256) = if source_version == source.active_version {
        (
            source.chunks.unwrap_or(0),
            source.size.unwrap_or(0),
            source.sha256.clone(),
        )
    } else {
        let v = Repository::new(&context.db)
            .versions(claims.sub)
            .get(source_file_id, source_version)
            .await?;
        (v.chunks, v.size, v.sha256)
    };

    // The client builds the new file's encrypted metadata; the server
    // overwrites the chunk-content fields with the source's recorded
    // values so the new row describes what's actually on disk.
    let mut payload = data.into_inner().file;
    payload.chunks = Some(chunks);
    payload.size = Some(size);
    payload.sha256 = sha256;

    let (mut active_model, encrypted_key, _hashed_tokens, _size, _parent) =
        payload.into_active_model()?;
    // Skip the chunk-upload pass — chunks land via copy_version below.
    let now = Utc::now().timestamp();
    active_model.chunks_stored = ActiveValue::Set(Some(chunks));
    active_model.finished_upload_at = ActiveValue::Set(Some(now));

    let storage = Fs::new(&context.config);

    let txn = context.db.begin().await?;
    let outcome = Repository::new(&txn)
        .versions(claims.sub)
        .fork(source_file_id, source_version, active_model, encrypted_key)
        .await?;
    txn.commit().await?;

    let new_file = Repository::new(&context.db)
        .manage(claims.sub)
        .file(outcome.new_file_id)
        .await?;

    // Source chunks live under the original file's path; destination
    // is the brand-new file's v1.
    storage
        .copy_version(
            &source,
            outcome.source_version,
            &new_file,
            new_file.active_version,
        )
        .await?;

    Ok(HttpResponse::Ok().json(new_file))
}

/// `DELETE /api/storage/{file_id}/versions/{version}` — drop a single
/// historical version. Deleting the active version is rejected; use
/// regular file deletion for that.
#[route(
    "/api/storage/{file_id}/versions/{version}",
    method = "DELETE"
)]
pub(crate) async fn delete(
    req: HttpRequest,
    claims: Claims,
    context: web::Data<Context>,
) -> AppResult<HttpResponse> {
    let context = context.into_inner();
    let file_id: Uuid = util::actix::path_var(&req, "file_id")?;
    let version: i32 = util::actix::path_var(&req, "version")?;

    Repository::new(&context.db)
        .versions(claims.sub)
        .delete(file_id, version)
        .await?;

    // On-disk cleanup is best-effort — DB row already gone, so an
    // orphaned dir gets caught by the next purge_all on file delete.
    let file = Repository::new(&context.db)
        .manage(claims.sub)
        .file(file_id)
        .await?;
    if let Err(e) = Fs::new(&context.config).purge_version(&file, version).await {
        log::warn!(
            "Failed to purge v{} dir for file {}: {}",
            version,
            file_id,
            e
        );
    }

    Ok(HttpResponse::NoContent().finish())
}

/// `DELETE /api/storage/{file_id}/versions` — drop every historical
/// version, keeping only the active one.
#[route("/api/storage/{file_id}/versions", method = "DELETE")]
pub(crate) async fn purge_all_history(
    req: HttpRequest,
    claims: Claims,
    context: web::Data<Context>,
) -> AppResult<HttpResponse> {
    let context = context.into_inner();
    let file_id: Uuid = util::actix::path_var(&req, "file_id")?;

    let pruned = Repository::new(&context.db)
        .versions(claims.sub)
        .purge_all_history(file_id)
        .await?;

    let file = Repository::new(&context.db)
        .manage(claims.sub)
        .file(file_id)
        .await?;
    let storage = Fs::new(&context.config);
    for version in pruned {
        if let Err(e) = storage.purge_version(&file, version).await {
            log::warn!(
                "Failed to purge v{} dir for file {}: {}",
                version,
                file_id,
                e
            );
        }
    }

    Ok(HttpResponse::NoContent().finish())
}

File: web/services/storage/versions.ts
/**
 * Version-history endpoints for editable files. Mirrors the server's
 * `/api/storage/{file_id}/versions/...` surface added in the
 * versioned-chunks atomicity work.
 *
 * All operations are owner-only. Restore and fork require a transfer
 * token-less call because the heavy chunk copy happens server-side.
 */

import Api from '../api'
import { evictChunkUrls, markDirectTransportBroken, versionChunkUrls } from './download/direct'
import type { AppFile, FileVersion } from 'types'

/**
 * Body the server's `fork` route expects. Mirrors `CreateFile` on the
 * server (every field optional) but we narrow to the ones the client
 * actually sets. `chunks`/`size`/`sha256` are ignored — the server
 * uses the source version's recorded values.
 */
export interface ForkRequest {
  name_hash: string
  encrypted_name: string
  encrypted_thumbnail?: string
  encrypted_key: string
  mime: string
  cipher: string
  editable?: boolean
  file_id?: string
  search_tokens_root?: string[]

  search_tokens_file?: string[]
}

/**
 * `GET /api/storage/{fileId}/versions` — newest-first list of historical
 * snapshots. The active version is intentionally absent (it lives on
 * the file row).
 */
export async function list(fileId: string): Promise<FileVersion[]> {
  const response = await Api.get<FileVersion[]>(`/api/storage/${fileId}/versions`)
  return response?.body || []
}

/**
 * Fetch a single encrypted chunk of a historical version. The caller decrypts
 * with the file's key.
 *
 * Goes straight to the storage bucket when the deployment serves presigned
 * URLs, and through `GET /api/storage/{fileId}/versions/{version}?chunk=N`
 * otherwise. Same manifest module every other read path uses, so a deployment
 * that transfers directly does so here too.
 */
export async function downloadChunk(
  fileId: string,
  version: number,
  chunk: number,
  signal?: AbortSignal
): Promise<Uint8Array> {
  const direct = (await versionChunkUrls(fileId, version))?.[chunk]

  // Credentials would oblige the bucket to answer with an exact-origin CORS
  // policy it does not carry, and are meaningless against a presigned URL.
  let response: Response | undefined
  try {
    response = direct ? await fetch(direct, { credentials: 'omit', signal }) : undefined
  } catch {
    // A broken CORS policy or an unreachable bucket throws instead of
    // answering; heal it the same way an error answer heals below, and stand
    // direct reads down for a while — no fresh manifest fixes a transport
    // that cannot be reached. Aborts rethrow from the relaying read, which
    // shares the signal.
    markDirectTransportBroken()
    response = undefined
  }

  if (direct && (!response || !response.ok)) {
    // An expired URL or a pruned object answers with an XML error document —
    // a body, but not the chunk. Read as ciphertext it would fail much later
    // as a baffling decrypt error, so drop the manifest and take the
    // relaying route for this read instead.
    evictChunkUrls(fileId)
    response = undefined
  }

  if (!response) {
    response = await new Api().download(
      `/api/storage/${fileId}/versions/${version}?chunk=${chunk}`,
      undefined,
      undefined,
      signal
    )
  }

  if (!response.body) {
    throw new Error(`Failed to download chunk ${chunk} of v${version}`)
  }

  const reader = response.body.getReader()
  let data = new Uint8Array(0)
  let done = false

  while (!done) {
    const chunkRead = await reader.read()
    if (chunkRead.value) {
      const merged = new Uint8Array(data.length + chunkRead.value.length)
      merged.set(data, 0)
      merged.set(chunkRead.value, data.length)
      data = merged
    }
    done = chunkRead.done
  }

  return data
}

/**
 * `POST /api/storage/{fileId}/versions/{version}/restore` — pointer
 * flip plus chunk copy server-side. Returns the file with the new
 * `active_version` slot already swapped in.
 */
export async function restore(fileId: string, version: number): Promise<AppFile> {
  const response = await Api.post<undefined, AppFile>(
    `/api/storage/${fileId}/versions/${version}/restore`
  )

  if (!response?.body?.id) {
    throw new Error(`Failed to restore v${version}`)
  }

  // The active version just moved, so any manifest held for this file points
  // at the version it moved away from.
  evictChunkUrls(fileId)

  return response.body
}

/**
 * `POST /api/storage/{fileId}/versions/{version}/fork` —
 * restore-as-new-note. The body is a `CreateFile` (same shape as
 * regular file creation): client builds the encrypted name/key/etc;
 * server copies the source's chunks into the new file's v1 and
 * overrides chunks/size/sha256 with the source version's recorded
 * values.
 */
export async function fork(
  fileId: string,
  version: number,
  newFile: ForkRequest
): Promise<AppFile> {
  const response = await Api.post<ForkRequest, AppFile>(
    `/api/storage/${fileId}/versions/${version}/fork`,
    undefined,
    newFile
  )

  if (!response?.body?.id) {
    throw new Error(`Failed to fork v${version}`)
  }

  return response.body
}

/**
 * `DELETE /api/storage/{fileId}/versions/{version}` — drop a single
 * historical snapshot. The active version cannot be deleted this way.
 */
export async function remove(fileId: string, version: number): Promise<void> {
  await Api.delete(`/api/storage/${fileId}/versions/${version}`)
}

/**
 * `DELETE /api/storage/{fileId}/versions` — wipe every historical
 * snapshot, keeping only the current active version.
 */
export async function purgeAll(fileId: string): Promise<void> {
  await Api.delete(`/api/storage/${fileId}/versions`)
}
